src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs

```diff
@@ -0,0 +1,146 @@
+            // at Microsoft.Spark.Sql.RowConstructor.GetRow() in Microsoft.Spark\Sql\RowConstructor.cs:line 113
+            // at Microsoft.Spark.Sql.RowCollector.Collect(ISocketWrapper socket) + MoveNext() in Microsoft.Spark\Sql\RowCollector.cs:line 36
+            // at Microsoft.Spark.Sql.DataFrame.GetRows(String funcName) + MoveNext() in Microsoft.Spark\Sql\DataFrame.cs:line 891
+            Func<Column, Column> udf = Udf<IDictionary<string, string>, string>(
+                    dict => dict.Count.ToString());
+            Assert.Throws<Exception>(() => _df.Select(udf(_df["info"])).Collect().ToArray());
+        }
+
+        /// <summary>
+        /// UDF that returns Map type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithReturnTypeAsMap()
+        {
+            // System.NotImplementedException: The method or operation is not implemented.
+            // at Microsoft.Spark.Sql.Row.Convert() in Microsoft.Spark\Sql\Row.cs:line 169
+            // at Microsoft.Spark.Sql.Row..ctor(Object[] values, StructType schema) in Microsoft.Spark\Sql\Row.cs:line 34
+            // at Microsoft.Spark.Sql.RowConstructor.GetRow() in MicrTosoft.Spark\Sql\RowConstructor.cs:line 113
+            // at Microsoft.Spark.Sql.RowCollector.Collect(ISocketWrapper socket) + MoveNext() in Microsoft.Spark\Sql\RowCollector.cs:line 36
+            // at Microsoft.Spark.Sql.DataFrame.GetRows(String funcName) + MoveNext() in Microsoft.Spark\Sql\DataFrame.cs:line 891
+            Func<Column, Column> udf = Udf<string, IDictionary<string, string>>(
+                str => new Dictionary<string, string> { { str, str } });
+            Assert.Throws<NotImplementedException>(() => _df.Select(udf(_df["name"])).Collect().ToArray());
+        }
+
+        /// <summary>
+        /// UDF that takes in Row type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithRowType()
+        {
+            Func<Column, Column> udf = Udf<Row, string>(
+                (row) =>
+                {
+                    string city = row.GetAs<string>("city");
+                    string state = row.GetAs<string>("state");
+                    return $"{city},{state}";
+                });
+
+            Row[] rows = _df.Select(udf(_df["info"])).Collect().ToArray();
+            Assert.Equal(3, rows.Length);
+
+            var expected = new[] { "Burdwan,Paschimbanga", "Los Angeles,California", "Seattle," };
+            for (int i = 0; i < rows.Length; ++i)
+            {
+                Row row = rows[i];
+                Assert.Equal(1, row.Size());
+                Assert.Equal(expected[i], row.GetAs<string>(0));
+            }
+        }
+
+        /// <summary>
+        /// UDF that returns Row type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithReturnTypeAsRow()
+        {
+            Assert.Throws<ArgumentException>(() => Udf<string, object[]>(

```

Comment:
I would put `Row` as the return type (also add exception messages)

Reply:
Done!